k2/python/k2/symbol_table.py

```diff
@@ -2,21 +2,38 @@
 #
 # See ../../../LICENSE for clarification regarding multiple authors
 
-from dataclasses import dataclass
-from typing import Dict
+from dataclasses import dataclass, field
+from typing import Dict, Optional
+from typing import Generic
+from typing import TypeVar
 from typing import Union
 
 
-@dataclass(frozen=True)
-class SymbolTable(object):
-    _id2sym: Dict[int, str]
+Symbol = TypeVar('Symbol')
+
+
+@dataclass(repr=False)  # Disable __repr__ otherwise it could freeze e.g. Jupyter.
+class SymbolTable(Generic[Symbol]):
+    '''SymbolTable that maps symbol IDs, found on the FSA arcs to
+    actual objects. These objects can be arbitrary Python objects
+    that can serve as keys in a dictionary (i.e. they need to be
+    hashable and immutable).
+
+    The SymbolTable can only be read to/written from disk if the
```

Comment:
Is this constraint enforced by the implementation?
I find that it uses `print` to convert a symbol to a string.

Reply:
It won’t know how to convert back to the correct type when reading. If this is needed in the future we could extend it I think.